pkg/querier/queryrange/roundtrip.go

```diff
@@ -236,14 +236,38 @@ func NewSeriesTripperware(
 	log log.Logger,
 	limits Limits,
 	codec queryrange.Codec,
+	extractor queryrange.Extractor,
 	instrumentMetrics *queryrange.InstrumentMiddlewareMetrics,
 	retryMiddlewareMetrics *queryrange.RetryMiddlewareMetrics,
 	splitByMetrics *SplitByMetrics,
-) (frontend.Tripperware, error) {
+) (frontend.Tripperware, Stopper, error) {
 	queryRangeMiddleware := []queryrange.Middleware{}
 	if cfg.SplitQueriesByInterval != 0 {
 		queryRangeMiddleware = append(queryRangeMiddleware, queryrange.InstrumentMiddleware("split_by_interval", instrumentMetrics), SplitByIntervalMiddleware(limits, codec, splitByMetrics))
 	}
+
+	var c cache.Cache
+	if cfg.CacheResults {
+		queryCacheMiddleware, cache, err := queryrange.NewResultsCacheMiddleware(
+			log,
+			cfg.ResultsCacheConfig,
+			cacheKeyLimits{limits},
+			limits,
+			codec,
+			extractor,
+			nil,
+		)
```

Comment:
Should we create separate `CacheMiddleware` for `series` or use common `CacheMiddleware` for both `series` and `metrics` ?

If we create separate `CacheMiddleware`,  I am not sure how to handle `stop` here https://github.com/grafana/loki/blob/268b7a3d0380b4621c862409ed42e95be8702fab/pkg/loki/modules.go#L316

Reply:
I covered this above, but you can make an adapter type that stops multiple caches.

Reply:
I'm actually not sure if it will work with the current cacheMiddleware it's possible.

I would wrote some test isolating only the middleware with couple of funky use cases and see what happens.

Reply:
Sure

Reply:
Well I have given some thoughts and I don't think you can use the current cache system for series. This is because within a given response you can't extract a subset of series since you don't know which series appears at what time.

This means you need to build your own cache system, but don't be afraid it's actually way much simpler that the cortex/query range one. You can only use a cache entry if the requested time range is wider than what you have in cache. Again you cannot extract a smaller subset from a cache entry.

Here is how I think it should work.
- for a given time range check if you have a cache entries (you can have multiple).
- if you don't just query the api and cache it storing the time range with it.
- if you do, you need to verify that the requested time range is bigger and that the current cache entry is within that time range (no overlap is allowed it can't overlap on another time range because again you can't extract a subset of series).
- use any possible entries and query missing entry.
- save all new and old entries.

Reply:
I should add that this new cache can be used for labels too.

Reply:
I was trying to add few tests to verify series API cache results. It fails in this test 
https://github.com/grafana/loki/blob/438bf456b0d1e4e527f523a22a6f20f0a55711f9/pkg/querier/queryrange/roundtrip_test.go#L325-L327

Build failure: https://cloud.drone.io/grafana/loki/3324/1/2

So, as you and @owen-d mentioned it won't return correct results for sub set time ranges.

First, I thought of just changing the `GenerateCacheKey` to consider `request.End` as well and everything else remains same. However, this will simply do a full request instead of a sub set request for large time ranges even when there is a subset data available in the cache. 

Anyhow, the steps you mentioned are very clear. Will implement the cache code accordingly.